### ISSUE TITLE
libscanmem: value.c: determining which uservalue_t flags to set

### ISSUE THREAD

**Issue by bkazemi** [scanmem/scanmem#234]
In https://github.com/scanmem/scanmem/blob/master/value.c#L209-L210 the int64_value and uint64_value members are unconditionally set which can cause undefined behavior/wrap-around when the numbers are out of range; thus, if a user supplied a number `x` where `x < INT32_MIN` there would be no reliable way to tell if the user indeed supplied a `int64_t` value or not, because both the signed and unsigned variables will be cast regardless. Before I opened a PR I wanted to confirm there was no subtle reason behind this.

**Comment by 12345ieee** [comment]
I'm not sure I follow you, any string that is correctly parsed by the `strtoll` a few lines over is a valid int64.
Were you referring to something else?

**Comment by bkazemi** [comment]
Yes, it is valid then, but when it gets cast to uint64 on line 209, that's where the problem is.

**Comment by 12345ieee** [comment]
Today I must not be very sharp, I keep not seeing it.
@bkazemi , could you please post your patch so I can figure out both the problem you saw and how you'd solve it?

These flags control the comparisons that will be made, you're pretty much forced to have the 64{s,u} flags set, otherwise it will never run any comparison.

**Comment by bkazemi** [comment]
```
--- a/value.c
+++ b/value.c
@@ -206,7 +206,7 @@ bool parse_uservalue_int(const char *nptr, uservalue_t * val)
     if (num >= -(1LL<<15) && num < (1LL<<15)) { val->flags.s16b = 1; set_s16b(val, (int16_t)num); }
     if (num >=        (0) && num < (1LL<<32)) { val->flags.u32b = 1; set_u32b(val, (uint32_t)num); }
     if (num >= -(1LL<<31) && num < (1LL<<31)) { val->flags.s32b = 1; set_s32b(val, (int32_t)num); }
-    if (           (true) &&          (true)) { val->flags.u64b = 1; set_u64b(val, (uint64_t)num); }
+    if (num >=        (0) &&          (true)) { val->flags.u64b = 1; set_u64b(val, (uint64_t)num); }
     if (           (true) &&          (true)) { val->flags.s64b = 1; set_s64b(val, (int64_t)num); }
 
     return true;
```

We simply check if `num` was non-negative before setting `u64b`. There are other issues, but this was the one I was addressing specifically.

**Comment by 12345ieee** [comment]
It sorta depends on what your intention is.

The current code basically "works on everything you give it".
Your code correctly tells me `-A` is not unsigned, but has lost the ability to work with numbers between 2^63 and 2^64 (beware of off-by-ones) as unsigned values.
Though admittedly even the current code makes it difficult because you need to input these numbers as the corresponding negative integers.

**Comment by bkazemi** [comment]
The problem is that the current code incorrectly handles the numbers. I was aware of the 2^64-2^63 precision loss, however, that is actually already lost due to the fact that `num` is of type `long long`. Casting to `uint64_t` does not make a difference at that point; the number has already been improperly converted to `long long`. We'd need to manually check for a negative sign and parse the number with `strtoll()` or `strtoull()` accordingly.

**Comment by sriemer** [comment]
Indeed, parsing for a sign would be safer to differentiate between int64_t and uint64_t. I'll build a test case for this. `0xdeadbeefdeadbeef` might be hidden in memory and can be used for both - signed and unsigned.

**Comment by 12345ieee** [comment]
What happened with this one? Now I definitely get the point.

The safest way is definitely @sriemer 's, via manual sign parsing and then a `uint64_t res = strtoull(...)` for the number.
Rebuilding the number later is easy once you've ensured it would be a valid `int64_t`.

**Comment by bkazemi** [comment]
I have so many things in the pipeline, I forgot about opening a PR. I'm working on some `scanmem` stuff today, I can open a quick PR.

**Comment by 12345ieee** [comment]
Again, no hurry, we're all doing this for fun. I thought you had it ready, I can do it as well.

I'm trying to wrap up all these low hanging fruits and release v0.17 before the summer period ends, at which point I'll have much less free time.

**Comment by 12345ieee** [comment]
I should have implemented this correctly in 630a1db .
Please ignore the different syntax for setting flags, I'm cooking a flags rehaul.
